src/ConfigProvider.php

```diff
@@ -58,11 +59,15 @@ public function getDependencies() : array
                 // AuthenticationInterface::class => Basic\BasicAccess::class,
                 // Provide an alias for the UserRepository adapter based on your application needs.
                 // UserRepositoryInterface::class => UserRepository\Htpasswd::class
+                UserInterface::class => DefaultUser::class
             ],
             'factories' => [
                 AuthenticationMiddleware::class => AuthenticationMiddlewareFactory::class,
                 UserRepository\Htpasswd::class => UserRepository\HtpasswdFactory::class,
                 UserRepository\PdoDatabase::class => UserRepository\PdoDatabaseFactory::class
+            ],
+            'invokables' => [
+                DefaultUser::class => DefaultUser::class
```

Comment:
What you and I had discussed earlier was that the user "service" would map to a _factory class_ that the authentication middleware would then compose; by doing that, we both de-couple how the instance is produced as well as the type produced.

I wonder if that might answer some of the questions from @xtreamwayz , as individual adapters could provide their own user implementations **and** the factories for producing them.

Reply:
It's a step closer to a solution. However, it's not only about a User object. It's also about storing in and retrieving from a session.